lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,404 @@
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
+        relation.data[self.charm.app]["tac"] = str(tac)
+        relation.data[self.charm.app].update({"plmns": json.dumps(plmns)})
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Triggered whenever a requirer charm joins the relation.
+
+        Args:
+            event (RelationJoinedEvent): Juju event
+        """
+        self.on.fiveg_core_gnb_request.emit(relation_id=event.relation.id)
+
+
+class FivegCoreGnbAvailableEvent(EventBase):
+    """Dataclass for the `fiveg_core_gnb` available event."""
+
+    def __init__(self, handle: Handle, tac: str, plmns: list[PLMNConfig]):
+        """Set CU / gNodeB's TAC and PLMNs."""
+        super().__init__(handle)
+        self.tac = tac
+        self.plmns = plmns
+
+    def snapshot(self) -> dict:
+        """Return event data."""
+        return {
+            "tac": self.tac,
+            "plmns": self.plmns,
+        }
+
+    def restore(self, snapshot: dict) -> None:
+        """Restore event data.
+
+        Args:
+            snapshot (dict): contains information to be restored.
+        """
+        self.tac = snapshot["tac"]
+        self.plmns = snapshot["plmns"]
+
+
+class FivegCoreGnbRequirerAppData(BaseModel):
+    """Requirer application data for fiveg_core_gnb."""
+    cu_ci: str = Field(
+        description="CU/gNB unique identifier",
+        examples=["gnb001"],
+    )
+
+
+class RequirerSchema(DataBagSchema):
+    """Requirer schema for fiveg_core_gnb."""
+
+    app_data: FivegCoreGnbRequirerAppData
+
+
+def data_matches_requirer_schema(data: dict) -> bool:
+    """Return whether data matches requirer schema.
+
+    Args:
+        data (dict): Data to be validated.
+
+    Returns:
+        bool: True if data matches requirer schema, False otherwise.
+    """
+    try:
+        RequirerSchema(app_data=FivegCoreGnbRequirerAppData(**data))
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
+
+class FivegCoreGnbRequirerCharmEvents(CharmEvents):
+    """Custom events for the FivegCoreGnbRequirer."""
+
+    fiveg_core_gnb_available = EventSource(FivegCoreGnbAvailableEvent)
+
+
+class FivegCoreGnbRequires(Object):
+    """Class to be instantiated by requirer of the `fiveg_core_gnb`."""
+
+    on = FivegCoreGnbRequirerCharmEvents()  # type: ignore
+
+    def __init__(self, charm: CharmBase, relation_name: str, cu_name: str):
+        """Observes relation joined and relation changed events.
+
+        Args:
+            charm: Juju charm
+            relation_name (str): Relation name
+            cu_name (str): name of the CU/gNodeB
+        """
+        self.relation_name = relation_name
+        self.cu_name = cu_name
+        self.charm = charm
+        super().__init__(charm, relation_name)
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_changed)
```

Comment:
Probably no need to handle this. When gNB joins the relation, all we need to do is publishing its name.